src/fusion/trash.nim

```
@@ -0,0 +1,84 @@
+  var fname = extractFilename(fullPath)
+  result =
+    when defined(linux) or defined(bsd): trashPath / "files" / fname
+    else: trashPath / fname
+  # If file exists on Trash, append " (1)", " (2)", " (3)", etc.
+  if fileExists(result):
+    for i in postfixStart .. int.high:
+      var prefix = " ("
+      prefix.add $i
+      prefix.add ')'
+      if not fileExists(result & prefix):
+        fname = fname & prefix
+        result =
+          when defined(linux) or defined(bsd): trashPath / "files" / fname
+          else: trashPath / fname
+        break
+  when defined(linux) or defined(bsd):
+    var trashinfo = "[Trash Info]\nPath="
+    trashinfo.add fullPath
+    trashinfo.add "\nDeletionDate="
+    trashinfo.add now().format("yyyy-MM-dd'T'HH:MM:ss")
+    trashinfo.add '\n'
+    discard existsOrCreateDir(trashPath / "files")
+    discard existsOrCreateDir(trashPath / "info")
+    moveFile(fullPath, result)
+    writeFile(trashPath / "info" / fname & ".trashinfo", trashinfo)
+  else:
+    moveFile(expandFilename(filename), result)
+
+
+proc moveFileFromTrash*(filename: string; trashPath = getTrash()) =
+  ## Move file from `trashPath` to `filename`, `trashPath` defaults to `getTrash()`.
+  ##
+  ## If a file with the same name already exists in the Trash folder,
+  ## then the generated postfix like `" (1)"`, `" (2)"`, `" (3)"`, etc,
+  ## is required on `filename` to restore the correct file.
+  ##
+  ## .. code-block:: nim
+  ##   import os, trash
+  ##   writeFile("example.txt", "")
+  ##   let trashedFile = moveFileToTrash("example.txt")
+  ##   moveFileFromTrash(getCurrentDir() / extractFilename(trashedFile))
+  assert filename.len > 0, "filename must not be empty string"
+  if dirExists(trashPath):
+    when defined(linux) or defined(bsd):
+      let fname = extractFilename(filename)
+      moveFile(trashPath / "files" / fname, filename)
+      discard tryRemoveFile(trashPath / "info" / fname & ".trashinfo")
```

why not `removeFile`, so we don't silently ignore errors? (note that `removeFile` won't raise if file doesn't exist)